.github/workflows/memcheck.yml

```diff
@@ -0,0 +1,150 @@
+        run: echo "Peer is $SAFE_PEERS"
+
+      # Start a heaptracked node instance to compare memory usage
+      - name: Start safenode with heaptrack
+        run: |
+          mkdir -p ~/.safe/heapnode
+          heaptrack ./target/release/safenode --root-dir ~/.safe/heapnode --log-dir ~/.safe/heapnode &
+          sleep 10
+
+      
+      - name: Start a client to upload files
+        run: cargo run --bin safe --release -- files upload -- "./resources"
+        id: client-file-upload
+        env:
+          RUST_LOG: "safenode,safe=trace"
+        timeout-minutes: 2
+
+      - name: Start a client to download files
+        run: cargo run --bin safe --release -- files download
+        id: client-file-download
+        env:
+          RUST_LOG: "safenode,safe=trace"
+        timeout-minutes: 2
+
+      - name: Start a client to create a register
+        run: cargo run --bin safe --release -- register create baobao
+        id: client-register-create
+        env:
+          RUST_LOG: "safenode,safe=trace"
+        timeout-minutes: 2
+
+      - name: Start a client to get a register
+        run: cargo run --bin safe --release -- register get baobao
+        id: client-register-get
+        env:
+          RUST_LOG: "safenode,safe=trace"
+        timeout-minutes: 2
+
+      - name: Start a client to edit a register
+        run: cargo run --bin safe --release -- register edit baobao wood
+        id: client-register-edit
+        env:
+          RUST_LOG: "safenode,safe=trace"
+        timeout-minutes: 2
+
+
+      - name: Kill all nodes
+        shell: bash
+        timeout-minutes: 1
+        continue-on-error: true
+        run: |
+          pkill safenode
+          echo "$(pgrep safenode | wc -l) nodes still running"
+
+      - name: Check for heaptrack file
+        run: ls -la
+
+      - name: Check memory usage
+        shell: bash
+        env:
+          MEM_LIMIT_MB: 10
+
+        run: |
+          HEAPTRACK_FILE=$(ls -t heaptrack.safenode.*.zst | head -1)
+          heaptrack --analyze $HEAPTRACK_FILE
+          MEMORY_USAGE=$(grep "peak: " "${HEAPTRACK_FILE%.zst}.txt" | awk '{print $2}')
+          if (( $(echo "$MEMORY_USAGE > $MEM_LIMIT_MB" | bc -l) )); then
+            echo "Memory usage exceeded threshold: $MEMORY_USAGE MB"
+            exit 1
+          fi
+
+      - name: Tar log files
+        shell: bash
+        continue-on-error: true
+        run: find ~/.safe/node/local-test-network -iname '*.log*' -o $/tmp/*.log | tar -zcvf log_files.tar.gz --files-from -
+        if: failure()
+
+      - name: Upload Node Logs
+        uses: actions/upload-artifact@main
+        with:
+          name: sn_node_logs_e2e_${{matrix.os}}
+          path: log_files.tar.gz
+        if: failure()
+        continue-on-error: true
+
+
+      - name: Upload Heaptrack
+        uses: actions/upload-artifact@main
+        with:
+          name: heaptrack_safenode
+          path: heaptrack.safenode.*
+        continue-on-error: true
```

Comment:
need an extra line break at the end of the file to get rid of the `red stop icon`